### PR TITLE
Store OIDC tokens

### PR DESCRIPTION
**Pull request by glombiewski**
- [ ] I added an entry to [`CHANGELOG.md`](CHANGELOG.md) if knowledge of this change could be valuable to users.

---

Here is a brief summary of what I did:

- Added AES password encryption for strings
- Added OIDC manager abstraction. An OIDC manager is always present, and either works or does not. This simplifies the Option<Arc<...>> handling of the previous top level OIDC abstraction.
- Stored access and refresh tokens in the database (encrypted if a password is in the config file).
- Use refresh tokens whenever a SessionContext is requested to extend the session if it has expired.

### CODE REVIEW

**Comment by glombiewski** [comment]
[Speicherung und Benutzung von OIDC Tokens](https://trello.com/c/wLmXIBIy/1666-speicherung-und-benutzung-von-oidc-tokens)

File: services/src/pro/contexts/migrations/current_schema.sql [hunk 1]
```diff
@@ -216,3 +216,12 @@ CREATE TABLE ml_models (
     id uuid PRIMARY KEY,
     content text NOT NULL
 );
+
+CREATE TABLE oidc_session_tokens (
+    session_id uuid REFERENCES sessions (id) ON DELETE CASCADE NOT NULL,
```
Comment (michaelmattig) [inline services/src/pro/contexts/migrations/current_schema.sql:221]: add unique constraint? or can there be multiple tokens per session?

File: services/src/pro/contexts/migrations/migration_0000_initial.sql [hunk 2]
```diff
@@ -211,3 +211,12 @@ CREATE TABLE ml_models (
     id uuid PRIMARY KEY,
     content text NOT NULL
 );
+
+CREATE TABLE oidc_session_tokens (
```
Comment (michaelmattig) [inline services/src/pro/contexts/migrations/migration_0000_initial.sql:215]: why did you add it to the initial migration and not a new one?

File: services/src/pro/users/oidc.rs [hunk 3]
```diff
@@ -359,29 +534,27 @@ impl OidcRequestDb {
...
-            issuer: value.issuer.to_string(),
-            client_id: value.client_id.to_string(),
-            client_secret: value.client_secret.clone(),
-            redirect_uri: value.redirect_uri.to_string(),
-            scopes: value.scopes,
-            users: Arc::new(Default::default()),
-            state_function: || CsrfToken::new(SINGLE_STATE.to_string()),
-            nonce_function: || Nonce::new(SINGLE_NONCE.to_string()),
-        }
+    async fn refresh_token(
```
Comment (michaelmattig) [inline services/src/pro/users/oidc.rs:547]: shouldn't this be called `access_token` as it return an access token?

File: services/src/pro/users/postgres_userdb.rs [hunk 4]
```diff
@@ -425,6 +435,17 @@ where
...
             .map_err(|_error| error::Error::InvalidSession)?;
 
+        let valid_session: bool = row.get(7);
+
+        let valid_until = if valid_session {
+            row.get(4)
+        } else {
+            self.refresh_tokens(session, &tx)
+                .await
+                .map_err(|_error| Error::InvalidSession)?
```
Comment (michaelmattig) [inline services/src/pro/users/postgres_userdb.rs:445]: is there a more meaningful error here? When can the token refresh fail? if the refresh token is not valid anymore?

File: services/src/pro/users/postgres_userdb.rs [hunk 5]
```diff
@@ -449,12 +470,204 @@ where
...
+                        &flat_tokens.access_token_nonce,
+                        &(oidc_tokens.expires_in.num_seconds() as f64),
+                        &flat_tokens.refresh_token_value,
+                        &flat_tokens.refresh_token_nonce,
+                    ],
+                )
+                .await?;
+
+            if updated_token_rows != 1 {
+                return Err(Error::InvalidSession);
```
Comment (michaelmattig) [inline services/src/pro/users/postgres_userdb.rs:601]: when can this happen and is there a more meaningful error (to log at least)?

File: services/src/pro/users/postgres_userdb.rs [hunk 6]
```diff
@@ -405,18 +413,20 @@ where
...
-                s.view
-            FROM 
-                sessions s JOIN user_sessions us ON (s.id = us.session_id) 
+                s.view,
+                CASE WHEN CURRENT_TIMESTAMP < us.valid_until THEN TRUE ELSE FALSE END AS valid_session
+            FROM
+                sessions s JOIN user_sessions us ON (s.id = us.session_id)
                     JOIN users u ON (us.user_id = u.id)
-            WHERE s.id = $1 AND CURRENT_TIMESTAMP < us.valid_until;",
+                    LEFT JOIN oidc_session_tokens t ON (s.id = t.session_id)
```
Comment (michaelmattig) [inline services/src/pro/users/postgres_userdb.rs:428]: would it make sense to first check whether the session is an oidc session and then have a separate logic for the token refresh that is separated from the local login?

File: services/src/pro/users/postgres_userdb.rs [hunk 7]
```diff
@@ -425,6 +435,17 @@ where
             .await
             .map_err(|_error| error::Error::InvalidSession)?;
 
+        let valid_session: bool = row.get(7);
+
+        let valid_until = if valid_session {
+            row.get(4)
+        } else {
+            self.refresh_tokens(session, &tx)
```
Comment (michaelmattig) [inline services/src/pro/users/postgres_userdb.rs:443]: don't you have to check whether the refresh token is still valid or where does this happen?

File: services/src/pro/users/oidc.rs (hunk 3 above)
Comment (glombiewski) [reply services/src/pro/users/oidc.rs]: Correct. I think refresh_access_token describes the entire process. ee8a2fc34b0071a07bfcd86d003e9da7adb57ff2

File: services/src/pro/users/postgres_userdb.rs (hunk 5 above)
Comment (glombiewski) [reply services/src/pro/users/postgres_userdb.rs]: Since the same transactions first checks if this row exists, I don't think this can occur. ee8a2fc34b0071a07bfcd86d003e9da7adb57ff2

File: services/src/pro/users/postgres_userdb.rs (hunk 4 above)
Comment (glombiewski) [reply services/src/pro/users/postgres_userdb.rs]: ee8a2fc34b0071a07bfcd86d003e9da7adb57ff2

I put a log debug message with the original error. My intention of mapping the error was that session_by_id call either gets a session or returns InvalidSession to whomever calls the function. I believe most API handlers use this as a first step, and they probably shouldn't care or propagate to the user that some internal refresh mechanism failed? But you are right, it should be noted somewhere, so I opted for a log for now.

When can this fail: 
The token refresh can fail if the token is not valid anymore. Or if the client does not respond for any reason. The other call to refresh_tokens is from get_access_token, where there might be no refresh_token (InvalidSession).

File: services/src/pro/util/config.rs [hunk 8]
```diff
@@ -37,14 +37,15 @@ pub enum QuotaTrackingMode {
...
-#[derive(Debug, Deserialize)]
+#[derive(Clone, Debug, Deserialize)]
 pub struct Oidc {
     pub enabled: bool,
     pub issuer: String,
     pub client_id: String,
     pub client_secret: Option<String>,
     pub redirect_uri: String, //TODO: Maybe URL type
     pub scopes: Vec<String>,
+    pub postgres_password: Option<String>,
```
[the remark this answers is not in the document]
Comment (glombiewski) [reply services/src/pro/util/config.rs]: Correct. 42eaf79b23a01e1176ac8ccaa39472d4ee46f9db

File: services/src/pro/users/postgres_userdb.rs (hunk 6 above)
Comment (glombiewski) [reply services/src/pro/users/postgres_userdb.rs:430]: We discussed this and opted to keep it as is.

File: services/src/pro/users/postgres_userdb.rs (hunk 7 above)
Comment (glombiewski) [reply services/src/pro/users/postgres_userdb.rs]: The Identity Provider does not send a expiration date with the refresh token. The refresh token might be invalid for a variety of reasons (expired, revoked, ...). The only way to check is to send it to the Identity Provider.

File: services/src/pro/contexts/migrations/current_schema.sql (hunk 1 above)
Comment (glombiewski) [reply services/src/pro/contexts/migrations/current_schema.sql]: Added a Primary Key, or would Unique be preferable for some reason? c09d87f913f283a000cff7c3c61c9aefb288fd1c

**Review by michaelmattig — APPROVED** [review]
